### ISSUE TITLE
TVIVideoView remote view is blank

### ISSUE THREAD

**Issue by vidhyasri2106** [twilio/video-quickstart-ios#384]
### Description

I have integrated programmable video in my sample app
I have 2 apps and both joins the same room.
On the publishing side it works just fine. I am able to view the preview of the other person and audio also works fine.
But on the receiver side, the remote view seems to be blank. It doesn't load the other person's back camera view itself.

Please find the swift file that has the receiver code in the below url,
https://www.dropbox.com/s/j0uxt3cv5iqznc0/ARHelpViewController.swift?dl=0

**Comment by paynerc** [comment]
@vidhyasri2106,

I have reviewed the code provided and at first glance it appears to be OK. A few questions for you:

1. Are you getting into the `func subscribed(to videoTrack: TVIRemoteVideoTrack, publication: TVIRemoteVideoTrackPublication, for participant: TVIRemoteParticipant)` callback and is the `setupRemoteViewView()` method being called, and is the `TVIVideoView` that is being created being displayed?
2. Would you be able to provide a RoomSID so I can review what's happening on our side? The best way to do that would be to send an email to `support@twilio.com` and reference this Github issue and that I asked you to provide that info.
3. Are both participants running this iOS code?
4. What version of the iOS SDK are you running?

Thank you,

Ryan

**Comment by vidhyasri2106** [comment]
Please find my answers below,

  1.  Yes the subscribed method is called and setupRemoteViewView()is also getting called but I couldn’t see any video rendered on screen. Its only blank. It’s strange because the same code was working a week back.
  2.  How do I get the room sid?
  3.  Yes both participants running in iOS code
  4.  I am using the below version

  pod 'TwilioVideo', '2.0.0-preview5'



Regards,
Vidhya Sri
Product Developer – Channel Apps Digital & Analytics
HCL Technologies Ltd.,
602/3, ELCOT, SEZ Unit – 1, Sholinganallur,
Chennai – 600 119
Direct: +91 9677523250

https://www.hcltech.com/digital-analytics-services<https://apac01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwww.hcltech.com%2Fdigital-analytics-services&data=02%7C01%7Cvidhyasri.so%40hcl.com%7Cded1b32f653f478f30d108d5a386b5e5%7C189de737c93a4f5a8b686f4ca9941912%7C0%7C0%7C636594721364915827&sdata=vPWPmJ8otwLyTlRPb5r2UGD%2F9%2FAuwESf%2F15p9gP50lQ%3D&reserved=0>

[id:image001.png@01D31687.A1963C50]

**Comment by paynerc** [comment]
Thanks for the quick reply!

My first suggestion is to update the version of Twilio Video you are using. `2.0.0-preview5` is very old and was before `2.0.0` went GA. The latest version of is `2.10.0`. You can update the line in your `Podfile` to read

```
  pod 'TwilioVideo', '~> 2.10'
```

and then run `pod update`.

Let me know if that doesn't immediately solve your issue.

Thanks!

Ryan

**Comment by vidhyasri2106** [comment]
I updated the pod and I am getting lots of compiler error like Cannot convert value of type 'Int64' to expected argument type 'CMTime'. Is there a sample for Programmable video sample for latest version of SDK?



Regards,
Vidhya Sri
Product Developer – Channel Apps Digital & Analytics
HCL Technologies Ltd.,
602/3, ELCOT, SEZ Unit – 1, Sholinganallur,
Chennai – 600 119
Direct: +91 9677523250

https://www.hcltech.com/digital-analytics-services<https://apac01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwww.hcltech.com%2Fdigital-analytics-services&data=02%7C01%7Cvidhyasri.so%40hcl.com%7Cded1b32f653f478f30d108d5a386b5e5%7C189de737c93a4f5a8b686f4ca9941912%7C0%7C0%7C636594721364915827&sdata=vPWPmJ8otwLyTlRPb5r2UGD%2F9%2FAuwESf%2F15p9gP50lQ%3D&reserved=0>

[id:image001.png@01D31687.A1963C50]

**Comment by paynerc** [comment]
Yes indeed, all of the examples in this repo, https://github.com/twilio/video-quickstart-swift, have been tested with `2.10.0`. You may need to do a git update to get the latest versions, but they will all compile fine without warnings and errors.

Ryan

**Comment by krishnads** [comment]
Hi, 

I am facing the same issue. Did you find the solution?

I am using 2.10.0 version.

**Comment by yaalisri** [comment]
> Hi,
> 
> I am facing the same issue. Did you find the solution?
> 
> I am using 2.10.0 version.

No. I upgraded it to the latest SDK and the issue was solved for sometime. But now, i am facing the same issue. Although the issue is not persistent. It occurs half of the time. Had to kill the app restart again to get it fixed. Are you facing the issue always or sometimes?

**Comment by krishnads** [comment]
> > Hi,
> > I am facing the same issue. Did you find the solution?
> > I am using 2.10.0 version.
> 
> No. I upgraded it to the latest SDK and the issue was solved for some time. But now, i am facing the same issue again. The issue is not persistent, it occurs half of the time. Had to kill the app and restart again to get it fixed. Are you facing the issue always or sometimes?

Yes, i am facing same issue everytime.

**Comment by FBS96** [comment]
@ceaglest have you found a solution to this problem?

**Comment by a1nouru** [comment]
I am facing the same issue. Did anyone find a solution for this?

**Comment by ceaglest** [comment]
Hi @a1nouru, @FBS96, @krishnads,

The QuickStart example only renders one RemoteVideoTrack at a time. However, the problem is that it doesn't select a new RemoteVideoTrack to render when the currently rendered track is unsubscribed. This logic is important:

```
    func didUnsubscribeFromVideoTrack(videoTrack: RemoteVideoTrack, publication: RemoteVideoTrackPublication, participant: RemoteParticipant) {
        // We are unsubscribed from the remote Participant's video Track. We will no longer receive the
        // remote Participant's video.
        
        logMessage(messageText: "Unsubscribed from \(publication.trackName) video track for Participant \(participant.identity)")

        // This logic is problematic when the LocalParticipant has already subscribed to a new RemoteVideoTrack.
        if (self.remoteParticipant == participant) {
            videoTrack.removeRenderer(self.remoteView!)
            self.remoteView?.removeFromSuperview()
            self.remoteView = nil

            // Now find a new Track to render, if any ...
        }
    }
```

I will update the example to fix the display bug.

Thanks for the feedback,
Chris

**Comment by a1nouru** [comment]
@ceaglest thanks a lot for the quick response. It does make sense. 
Looking forward to your update :) 

Thanks again.

**Comment by FBS96** [comment]
Thank you @ceaglest! I'm very excited for this update to happen

**Comment by ceaglest** [comment]
Hi @FBS96, @a1nouru, @krishnads,

I have posted a PR which updates the Quickstart example to be more flexible when rendering remote video. It still only renders 1 RemoteVideoTrack and 1 LocalVideoTrack at a time, but there should never be a case where 1 RemoteVideoTrack is subscribed and not being rendered by the LocalParticipant's app.

I've opened this PR https://github.com/twilio/video-quickstart-ios/pull/442. I will probably leave it in review for the next day or so. If you're interested, feel free to try the example and report back if it solves the problem for you.

Best,
Chris
